### PR TITLE
Add device context parameter

### PR DESCRIPTION
**Pull request by vlad-nazarov**

### CODE REVIEW

File: bench.py [hunk 1]
```diff
@@ -175,7 +175,11 @@ def parse_args(parser, size=None, loop_types=(),
                         help='Dataset name')
     parser.add_argument('--no-intel-optimized', default=False, action='store_true',
                         help='Use no intel optimized version. '
-                             'Now avalible for scikit-learn benchmarks'),
+                             'Now avalible for scikit-learn benchmarks')
+    parser.add_argument('--device', default=None, type=str,
```
Comment (PetrovKP) [inline bench.py:179]:
```suggestion
    parser.add_argument('--device', default="host", type=str,
```

File: configs/skl_with_context_config.json [hunk 2]
```diff
@@ -0,0 +1,77 @@
+{
+    "common": {
+        "lib": ["sklearn"],
+        "data-format": ["pandas"],
+        "data-order": ["F"],
+        "dtype": ["float64"],
+        "device": ["host", "cpu", "gpu"]
```
Comment (PetrovKP) [inline configs/skl_with_context_config.json:7]: What happens if I run this config on a machine without a GPU driver?

File: sklearn_bench/dbscan.py [hunk 3]
```diff
@@ -19,38 +19,41 @@
...
+
+    # Load generated data
+    X, _, _, _ = bench.load_data(params, add_dtype=True)
+
+    # Create our clustering object
+    dbscan = DBSCAN(eps=params.eps, n_jobs=params.n_jobs,
+                    min_samples=params.min_samples, metric='euclidean',
+                    algorithm='auto')
+
+    # N.B. algorithm='auto' will select DAAL's brute force method when running
```
Comment (PetrovKP) [inline sklearn_bench/dbscan.py:35]:
```suggestion
    # N.B. algorithm='auto' will select oneAPI Data Analytics Library (oneDAL) brute force method when running
```

**Review by PetrovKP — COMMENTED** [review]
overall good, but need to check the performance of the cpu (so that your changes do not bring an overhead). come to me

File: bench.py (hunk 1 above)
Comment (michael-smirnov) [reply bench.py:179]: My understanding is that `None` is used to run without context. Other values specify device type for a context

File: bench.py [hunk 4]
```diff
@@ -175,7 +175,11 @@ def parse_args(parser, size=None, loop_types=(),
...
+                        choices=("host", "cpu", "gpu"),
```
Comment (michael-smirnov) [inline bench.py:180]: `None` shall be also included?

File: bench.py [hunk 5]
```diff
@@ -197,6 +201,8 @@ def parse_args(parser, size=None, loop_types=(),
         except ImportError:
             print('Failed to import daal4py.sklearn.patch_sklearn.'
                   'Use stock version scikit-learn', file=sys.stderr)
+    else:
+        params.device = None
```
Comment (michael-smirnov) [inline bench.py:205]: I think we should check if the device parameter is passed by the user and print a warning that it is useless in that case - for clarity

File: configs/skl_with_context_config.json (hunk 2 above)
Comment (michael-smirnov) [reply configs/skl_with_context_config.json:7]: Exactly the same what happens if you try to run on a CPU wo DPC++ support - an exception. What is your suggession here?

Comment (michael-smirnov) [reply configs/skl_with_context_config.json:7]: We shall point somewhere that using this config file requires DPC++ support and GPU device on board

File: runner.py [hunk 6]
```diff
@@ -70,6 +70,9 @@ def generate_cases(params):
     parser.add_argument('--report', default=False, action='store_true',
                         help='Create an Excel report based on benchmarks results. '
                              'Need "openpyxl" library')
+    parser.add_argument('--device', default=None, type=str,
```
Comment (michael-smirnov) [inline runner.py:73]: Why the parameter is duplicated in `bench.py` and `runner.py`?

File: sklearn_bench/dbscan.py (hunk 3 above)
Comment (michael-smirnov) [reply sklearn_bench/dbscan.py:35]: @PetrovKP what about other files that @vlad-nazarov did not touch?

File: bench.py (hunk 1 above)
Comment (PetrovKP) [reply bench.py:179]: ohh, then I think the host is not needed at all

File: configs/skl_with_context_config.json (hunk 2 above)
Comment (PetrovKP) [reply configs/skl_with_context_config.json:7]: Ok, probably it's ok

File: sklearn_bench/dbscan.py (hunk 3 above)
Comment (PetrovKP) [reply sklearn_bench/dbscan.py:35]: I will correct if there are such places yet

File: bench.py [hunk 7]
```diff
@@ -197,6 +201,12 @@ def parse_args(parser, size=None, loop_types=(),
         except ImportError:
             print('Failed to import daal4py.sklearn.patch_sklearn.'
                   'Use stock version scikit-learn', file=sys.stderr)
+            params.device = 'host'
+    else:
+        if params.device != 'host':
+            print('Device context not supported without intel optimized version',
```
Comment (michael-smirnov) [inline bench.py:207]: Device context is not supported for stock scikit-learn. Please use `--no-intel-optimized=False` with f'--device={params.device}' parameter. Fallback to `--device=None`.

File: configs/skl_with_context_config.json [hunk 8]
```diff
@@ -0,0 +1,77 @@
+{
+    "common": {
+        "lib": ["sklearn"],
```
Comment (michael-smirnov) [inline configs/skl_with_context_config.json:3]: How stock or intel version of sk is specified for this config?

**Review by michael-smirnov — CHANGES_REQUESTED** [review]

**Review by PetrovKP — APPROVED** [review]
merge after green CI

File: configs/skl_with_context_config.json (hunk 8 above)
Comment (vlad-nazarov) [reply configs/skl_with_context_config.json:3]: Do we need to launch the stock sk in this config? Maybe just add flag `no-intel-optimized` before config run? I think it's better to separate patched and unpatched sk launches

Comment (michael-smirnov) [reply configs/skl_with_context_config.json:3]: Ok - probably its better to skip `device != None` branches for stock sklearn

**Review by michael-smirnov — APPROVED** [review]

File: configs/skl_with_context_config.json (hunk 8 above)
Comment (vlad-nazarov) [reply configs/skl_with_context_config.json:3]: Need to add support to skip these cases?

**Pull request merged**